onadata/libs/utils/middleware.py

```diff
@@ -49,6 +77,103 @@ def process_response(self, request, response):
         return response
 
 
+class RestrictedAccessMiddleware(MiddlewareMixin):
+
+    def __init__(self, get_response):
+        super().__init__(get_response)
+        self._allowed_view = None
+
+    def process_response(self, request, response):
+        if not request.user.is_authenticated:
+            return response
+
+        if isinstance(request.user, ServiceAccountUser):
+            return response
+
+        try:
+            profile = request.user.profile
+        except get_user_model().profile.RelatedObjectDoesNotExist:
+            # Consider user's password as weak
+            if not self._allowed_view:
+                return self._render_response(request, response)
+
+        if profile.validated_password:
+            return response
+
+        if not self._allowed_view:
+            return self._render_response(request, response)
+
+        return response
+
+    def process_view(self, request, view, view_args, view_kwargs):
+        """
+        Validate if the view is among the ones allowed to be accessed with
+        an unsafe password.
+        If it is not, set `self._allowed_view` to False to alter the
+        response in `process_response()`.
+
+        We cannot validate user's password here because DRF authentication
+        takes places after this method call. Thus, `request.user` is always
+        anonymous if user is authenticated with something else than the session.
+        """
+        view_name = view.__name__
+
+        # Reset boolean for each processed view
+        self._allowed_view = True
+
+        if request.method == 'HEAD':
+            return
+
+        try:
+            allowed_actions = ALLOWED_VIEWS_WITH_WEAK_PASSWORD[view_name][
+                request.method
+            ]
+        except KeyError:
+            self._allowed_view = False
+        else:
+            if hasattr(view, 'actions'):
+                view_action = view.actions[request.method.lower()]
+                if view_action not in allowed_actions:
+                    self._allowed_view = False
+
+        return
+
+    def _render_response(
+        self, request: HttpRequest, response: HttpResponse
+    ) -> Union[
+        HttpResponseForbidden, JsonResponseForbidden, XMLResponseForbidden
+    ]:
+        """
+        Render response in the requested format: HTML, JSON or XML.
+        If content type is not detected, fallback on HTML.
+        """
+        template = get_template('restricted_access.html')
+        format_ = None
+        try:
+            content_type, *_ = response.accepted_media_type.split(';')
+        except AttributeError:
+            pass
+        else:
+            *_, format_ = content_type.split('/')
```

Comment:
Learned something new! :mortar_board:
I think this way of considering only the first `accepted_media_type` isn't totally kosher, but it's fine for this purpose